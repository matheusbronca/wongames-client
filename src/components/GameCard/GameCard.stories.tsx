import { Meta, StoryObj } from '@storybook/react'
import GameCard from './GameCard'

const commonProps = {
  title: 'Game Title',
  publisher: 'Game publisher',
  listPrice: 200,
  sellingPrice: 100,
  imgUrl: '/img/red-dead-img.jpg',
  publisherUrl: '#',
  gameUrl: '/games/red-dead-redemption-2'
} satisfies Parameters<typeof GameCard>[0]

const meta = {
  title: 'GameCard',
  component: GameCard
} satisfies Meta<typeof GameCard>

export default meta

type Story = StoryObj<typeof meta>

export const Default = {
  args: commonProps,
  render: (args) => {
    return (
      <div className="bg-mainBg p-12 w-min">
        <GameCard {...args} className="w-72" />
      </div>
    )
  }
} satisfies Story

export const WithRibbon = {
  args: {
    ...commonProps,
    ribbonText: '20% OFF'
  },
  render: (args) => {
    return (
      <div className="bg-mainBg p-12 w-min">
        <GameCard {...args} className="w-72" />
      </div>
    )
  }
} satisfies Story
