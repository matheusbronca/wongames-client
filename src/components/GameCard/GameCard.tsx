import { cn } from '@/lib/utils'

import WishlistIcon from '@svg-icons/material-outlined/favorite-border.svg'
import WishlistFilledIcon from '@svg-icons/material-outlined/favorite.svg'
import AddToShoppingCartIcon from '@svg-icons/material-outlined/add-shopping-cart.svg'

import Ribbon from '../Ribbon/Ribbon'
import Button from '../Button/Button'

export type GameCardProps = {
  title: string
  publisher: string
  publisherUrl: string
  sellingPrice: number
  listPrice?: number
  imgUrl: string
  gameUrl: string
  ribbonText?: string
  className?: string
  isFavorite?: boolean
}

/* eslint-disable @next/next/no-img-element */
const GameCard = ({
  title,
  publisher,
  sellingPrice,
  listPrice,
  imgUrl,
  publisherUrl,
  gameUrl,
  ribbonText,
  className,
  isFavorite = false
}: GameCardProps) => {
  return (
    <article
      className={cn(
        'flex flex-col items-center w-full bg-white shadow-lg rounded rounded-t-none relative',
        className
      )}
    >
      {!!ribbonText && <Ribbon size="small">{ribbonText}</Ribbon>}
      <figure className="block w-full bg-gray">
        <a href={gameUrl} data-img-game-link>
          <img className="object-cover w-full h-36" src={imgUrl} alt={title} />
        </a>
        <div className="bg-white p-4 pt-3 pb-0 flex justify-between gap-4 items-start">
          <figcaption>
            <h3 className="font-semibold text-base">
              <a href={gameUrl} data-title-game-link>
                {title}
              </a>
            </h3>
            <a href={publisherUrl}>
              <address className="text-xs text-gray/60 not-italic hover:underline">
                {publisher}
              </address>
            </a>
          </figcaption>
          <button
            type="button"
            title="Add to wishlist"
            aria-label="Add to wishlist"
            className="size-6 text-primary"
            data-wishlist
          >
            {isFavorite ? <WishlistFilledIcon /> : <WishlistIcon />}
          </button>
        </div>
      </figure>
      <div className="flex items-center justify-end gap-1 size-full p-4 pt-2">
        {!!listPrice && (
          <div
            data-list-price
            className="text-gray/80 font-semibold line-through text-sm mr-3"
          >
            $
            {listPrice.toLocaleString('pt-BR', {
              maximumFractionDigits: 2,
              minimumFractionDigits: 2
            })}
          </div>
        )}
        <div
          data-selling-price
          className="text-sm font-semibold text-white rounded-md bg-secondary py-1 px-3"
        >
          $
          {sellingPrice.toLocaleString('pt-BR', {
            maximumFractionDigits: 2,
            minimumFractionDigits: 2
          })}
        </div>
        <Button Icon={<AddToShoppingCartIcon />} size="small" />
      </div>
    </article>
  )
}

export default GameCard
