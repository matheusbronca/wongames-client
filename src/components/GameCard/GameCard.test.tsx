import { render, screen } from '@testing-library/react'

import GameCard from './GameCard'

const commonProps = {
  title: 'Game Title',
  publisher: 'Game publisher',
  listPrice: 200,
  sellingPrice: 100,
  imgUrl: '/img/red-dead-img.jpg',
  publisherUrl: '#',
  gameUrl: '/games/red-dead-redemption-2'
} satisfies Parameters<typeof GameCard>[0]

describe('<GameCard />', () => {
  it('should be rendered correctly in the DOM', () => {
    const { container, debug } = render(<GameCard {...commonProps} />)

    debug()

    expect(screen.getByRole('article')).toBeInTheDocument()

    expect(screen.getByRole('img')).toHaveAttribute('src')
    expect(screen.getByRole('heading', { name: /Game Title/i }))
    expect(container.querySelector('address')).toHaveTextContent(
      /Game publisher/i
    )

    expect(
      screen.getByRole('button', { name: /Add to wishlist/i })
    ).toBeInTheDocument()
    expect(container.querySelector('[data-img-game-link]')).toHaveAttribute(
      'href',
      commonProps.gameUrl
    )

    expect(container.querySelector('[data-title-game-link]')).toHaveAttribute(
      'href',
      commonProps.gameUrl
    )

    expect(container.querySelector('[data-img-game-link]')).toHaveAttribute(
      'href',
      commonProps.gameUrl
    )

    expect(
      screen.getByRole('link', { name: /Game publisher/i })
    ).toHaveAttribute('href', commonProps.publisherUrl)

    expect(container.querySelector('[data-selling-price]')?.textContent).toBe(
      '$100,00'
    )

    expect(container.querySelector('[data-list-price]')?.textContent).toBe(
      '$200,00'
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the game card with a ribbon when passed in the props', () => {
    render(<GameCard {...commonProps} ribbonText="20% OFF" />)
    expect(screen.getByText(/20% OFF/i)).toBeInTheDocument()
  })

  it('should render the wishlist button with the filled icon, when the game is marked as favorite', () => {
    const { container } = render(<GameCard {...commonProps} isFavorite />)
    expect(container.querySelector('[data-wishlist] path')).toHaveAttribute(
      'd',
      'm12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z'
    )
  })
})
