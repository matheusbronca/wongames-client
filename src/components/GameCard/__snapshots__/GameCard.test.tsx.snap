// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameCard /> should be rendered correctly in the DOM 1`] = `
<article
  class="flex flex-col items-center w-full bg-white shadow-lg rounded rounded-t-none relative"
>
  <figure
    class="block w-full bg-gray"
  >
    <a
      data-img-game-link="true"
      href="/games/red-dead-redemption-2"
    >
      <img
        alt="Game Title"
        class="object-cover w-full h-36"
        src="/img/red-dead-img.jpg"
      />
    </a>
    <div
      class="bg-white p-4 pt-3 pb-0 flex justify-between gap-4 items-start"
    >
      <figcaption>
        <h3
          class="font-semibold text-base"
        >
          <a
            data-title-game-link="true"
            href="/games/red-dead-redemption-2"
          >
            Game Title
          </a>
        </h3>
        <a
          href="#"
        >
          <address
            class="text-xs text-gray/60 not-italic hover:underline"
          >
            Game publisher
          </address>
        </a>
      </figcaption>
      <button
        aria-label="Add to wishlist"
        class="size-6 text-primary"
        data-wishlist="true"
        title="Add to wishlist"
        type="button"
      >
        <svg-url />
      </button>
    </div>
  </figure>
  <div
    class="flex items-center justify-end gap-1 size-full p-4 pt-2"
  >
    <div
      class="text-gray/80 font-semibold line-through text-sm mr-3"
      data-list-price="true"
    >
      $
      200,00
    </div>
    <div
      class="text-sm font-semibold text-white rounded-md bg-secondary py-1 px-3"
      data-selling-price="true"
    >
      $
      100,00
    </div>
    <button
      class="text-white no-underline border-none rounded bg-gradient-to-b cursor-pointer from-[#FF5F5F] to-[#F062C0] hover:bg-gradient-to-b hover:from-[#e35565] hover:to-[#d958a6] text-xs h-7 p-2 inline-flex items-center justify-center [&_>_svg]:w-4 [&_svg_+_span]:ml-xxsm"
    >
      <svg-url />
    </button>
  </div>
</article>
`;
